lib/capistrano/configuration/scm_resolver.rb

```diff
@@ -0,0 +1,149 @@
+    #
+    # This SCMResolver class takes care of managing the transition from the old
+    # to new system. It maintains the legacy behavior, but prints deprecation
+    # warnings when it is used.
+    #
+    # To maintain backwards compatibility, the resolver will load the Git SCM by
+    # if default it determines that no SCM has been explicitly specified or
+    # loaded. To force no SCM to be used at all, use `set :scm, nil`. This hack
+    # won't be necessary once backwards compatibility is removed in a future
+    # version.
+    #
+    # TODO: Remove this class entirely in Capistrano 4.0.
+    #
+    class SCMResolver
+      DEFAULT_GIT = :"default-git"
+
+      include Capistrano::DSL
+
+      def resolve
+        return if scm_name.nil?
+        set(:scm, :git) if using_default_scm?
+
+        print_deprecation_warnings_if_applicable
+        return if scm_plugin_loaded?
+
+
+        if built_in_scm_name?
+          load_built_in_scm
+        else
+          # Compatibility with existing 3.x third-party SCMs
+          register_legacy_scm_hooks
+          load_legacy_scm_by_name
+        end
+      end
+
+      private
+
+      def using_default_scm?
+        return @using_default_scm if defined? @using_default_scm
+        @using_default_scm = (fetch(:scm) == DEFAULT_GIT)
+      end
+
+      # This is somewhat of a hack, because there is no guarantee that a third-
+      # party SCM will necessarily implement the typical SCM tasks. But it works
+      # well enough for the built-in SCMs.
+      def scm_plugin_loaded?
+        Rake::Task.tasks.any? { |t| t.name =~ /^[^:]+:create_release/ }
+      end
+
+      def scm_name
+        fetch(:scm)
+      end
+
+      def load_built_in_scm
+        require "capistrano/scm/#{scm_name}"
+        scm_class = Object.const_get(built_in_scm_plugin_class_name)
+        # We use :load_immediately because we are initializing the SCM plugin
+        # late in the load process and therefore can't use the standard
+        # load:defaults technique.
+        scm_class.new(:load_immediately => true)
+      end
+
+      def load_legacy_scm_by_name
+        load("capistrano/#{scm_name}.rb")
+      end
+
+      def third_party_scm_name?
+        !built_in_scm_name?
+      end
+
+      def built_in_scm_name?
+        %w(git hg svn).include?(scm_name.to_s.downcase)
+      end
+
+      def built_in_scm_plugin_class_name
+        "Capistrano::SCM::#{scm_name.to_s.capitalize}"
+      end
+
+      def register_legacy_scm_hooks
+        if Rake::Task.task_defined?("deploy:new_release_path")
+          after "deploy:new_release_path", "#{scm_name}:create_release"
+        end
+
+        if Rake::Task.task_defined?("deploy:check")
+          before "deploy:check", "#{scm_name}:check"
+        end
+
+        if Rake::Task.task_defined?("deploy:set_current_revision")
+          before "deploy:set_current_revision",
+                 "#{scm_name}:set_current_revision"
```

Comment:
This is simply replicating the existing behavior, which you can see in the master branch here:
https://github.com/capistrano/capistrano/blob/3d4b17c2fb10291d0a31adbd8b3cd3a086f3c9bf/lib/capistrano/tasks/deploy.rake#L228

In master, it assumed that if you use `set :scm, "foo"`, then `foo:create_release`, `foo:check`, and `foo:set_current_revision` *must* exist. I am maintaining that behavior here when the resolver uses "legacy mode".